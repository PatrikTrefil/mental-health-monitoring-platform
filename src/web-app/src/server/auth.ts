import {
    RequestError,
    fetchRoleList,
    loginAdmin,
    loginUser,
    refreshToken,
} from "@/client/formioClient";
import { UserFormSubmission } from "@/types/userFormSubmission";
import { UserRoleTitle, UserRoleTitles } from "@/types/users";
import { type GetServerSidePropsContext } from "next";
import { AuthOptions, DefaultSession, getServerSession } from "next-auth";
import CredentialsProvider from "next-auth/providers/credentials";
import { z } from "zod";

// make auto-completion work for useSession
declare module "next-auth" {
    interface User extends UserFormSubmission {
        roleTitles: UserRoleTitle[];
        formioToken: string;
        formioTokenExpiration: number;
    }
    interface Session {
        user: UserFormSubmission &
            DefaultSession["user"] & {
                roleTitles: UserRoleTitle[];
                formioToken: string;
            };
    }
}

declare module "next-auth/jwt" {
    interface JWT {
        user?: UserFormSubmission &
            DefaultSession["user"] & {
                roleTitles: UserRoleTitle[];
                formioToken: string;
                formioTokenExpiration: number;
            };
    }
}

/**
 * In miliseconds
 */
const formioTokenExpirationTime = 5 * 60 * 1000; // 5 minutes (default in formio)

export const authOptions: AuthOptions = {
    // Configure one or more authentication providers
    providers: [
        CredentialsProvider({
            name: "Credentials",
            /**
             * Used to create a login page and used as a param of the authorize function
             */
            credentials: {
                ID: { label: "ID", type: "text", placeholder: "1234" },
                password: { label: "Password", type: "password" },
            },
            /**
             * Called when initializing a session. The returned object will be passed to the jwt callback.
             * @returns Promise<null> if the credentials are invalid, otherwise Promise<User>
             */
            async authorize(credentials) {
                if (!credentials) return null; // invalid credentials because they were not provided

                try {
                    const { user, formioToken } = await loginUser(
                        credentials.ID,
                        credentials.password
                    );

                    const adminToken = await loginAdmin(
                        process.env.FORMIO_ROOT_EMAIL,
                        process.env.FORMIO_ROOT_PASSWORD
                    );

                    const roleList = await fetchRoleList(adminToken);

                    const roleTitles = user.roles.map((roleId) => {
                        const title = roleList.find(
                            (role) => role._id === roleId
                        )?.title;
                        return z.nativeEnum(UserRoleTitles).parse(title);
                    });

                    console.debug("Sending response...");
                    return {
                        id: user._id,
                        ...user,
                        formioToken,
                        roleTitles,
                        formioTokenExpiration:
                            Date.now() + formioTokenExpirationTime,
                    };
                } catch (e) {
                    console.error("Error: ", e); // log on server
                    if (e instanceof RequestError) {
                        const isInvalidCredentials =
                            e.statusCode >= 400 && e.statusCode < 500;
                        if (isInvalidCredentials) return null;
                    }
                    throw new Error("Unexpected error on server"); // Send generic error to client
                }
            },
        }),
    ],
    jwt: {
        maxAge: 230, // this must be lower that the validity of the formio token
    },
    callbacks: {
        /**
         * Creates the JWT token on the server.
         */
        async jwt({ token, user }) {
            // the user is available only in the first call
            // initial sign in
            if (user) {
                token.user = user;
            } else {
                if (!token.user) throw new Error("No user in token");
                // update the formio token (may throw)
                if (
                    Date.now() - token.user.formioTokenExpiration >
                    3 * 60 * 1000 // 3 minutes have passed
                ) {
                    const newToken = await refreshToken(token.user.formioToken);
                    token.user.formioToken = newToken;
                }
            }

            return token;
        },
        /**
         * Creates the session object on the server, which is sent to the client.
         */
        async session({ session, token }) {
            // make user information available in the session (used in client)
            if (token.user) session.user = token.user;

            return session;
        },
    },
    pages: {
        signIn: "/login",
    },
    debug:
        process.env.NODE_ENV === "development" ||
        process.env.NODE_ENV === "test",
};

/**
 * Wrapper for `getServerSession` so that you don't need to import the `authOptions` in every file.
 *
 * @see https://next-auth.js.org/configuration/nextjs
 */
export const getServerAuthSession = (ctx: {
    req: GetServerSidePropsContext["req"];
    res: GetServerSidePropsContext["res"];
}) => {
    return getServerSession(ctx.req, ctx.res, authOptions);
};
