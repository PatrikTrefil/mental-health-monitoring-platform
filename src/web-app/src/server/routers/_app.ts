import { z } from "zod";
import { createTRPCRouter, protectedProcedure, publicProcedure } from "../trpc";

export const appRouter = createTRPCRouter({
    hello: publicProcedure
        .input(
            z.object({
                text: z.string(),
            })
        )
        .query((opts) => {
            return {
                greeting: `hello ${opts.input.text}`,
            };
        }),
    helloUser: protectedProcedure.query((opts) => {
        return {
            greeting: `hello ${opts.ctx.session.user?.data.id}`,
        };
    }),
});

// export type definition of API
export type AppRouter = typeof appRouter;
