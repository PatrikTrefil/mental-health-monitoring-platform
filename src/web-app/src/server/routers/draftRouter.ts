import UserRoleTitles from "@/constants/userRoleTitles";
import { Prisma, type Draft } from "@prisma/client";
import { TRPCError } from "@trpc/server";
import { z } from "zod";
import { createTRPCRouter, protectedProcedure } from "../trpc";

const notFoundErrorCode = "P2025";

const draftDataSchema = z.record(z.string(), z.unknown());

const draftRouter = createTRPCRouter({
    /**
     * Update or insert a draft.
     * @param string - Input.formId the if of the form to update/insert a draft for.
     * @param unknown - Input.data the draft data.
     * @returns The updated/inserted draft.
     * @throws {TRPCError}
     * FORBIDDEN if the user is not a {@link UserRoleTitles.ASSIGNEE}.
     * @throws {TRPCError}
     * INTERNAL_SERVER_ERROR if the database operation fails.
     * @throws {TRPCError}
     * CUSTOM if the data can not be stringified to JSON.
     */
    upsertDraft: protectedProcedure
        .input(
            z.object({
                formId: z.string(),
                data: draftDataSchema.transform((data, ctx) => {
                    try {
                        return JSON.stringify(data);
                    } catch (e) {
                        ctx.addIssue({
                            message: "Data can not be stringified to JSON",
                            code: z.ZodIssueCode.custom,
                        });
                        return z.NEVER;
                    }
                }),
            })
        )
        .mutation(async (opts): Promise<Draft> => {
            if (
                !opts.ctx.session.user.roleTitles.includes(
                    UserRoleTitles.ASSIGNEE
                )
            )
                throw new TRPCError({ code: "FORBIDDEN" });

            try {
                return await opts.ctx.prisma.draft.upsert({
                    where: {
                        formId_userId: {
                            formId: opts.input.formId,
                            userId: opts.ctx.session.user.data.id,
                        },
                    },
                    update: {
                        data: opts.input.data,
                    },
                    create: {
                        formId: opts.input.formId,
                        userId: opts.ctx.session.user.data.id,
                        data: opts.input.data,
                    },
                });
            } catch (e) {
                console.error(e);
                throw new TRPCError({ code: "INTERNAL_SERVER_ERROR" });
            }
        }),
    /**
     * Get a draft by form id.
     * @param string - Input.formId the id of the form to get the draft for.
     * @returns The requested draft.
     * @throws {TRPCError}
     * FORBIDDEN if the user is not a {@link UserRoleTitles.ASSIGNEE}.
     * @throws {TRPCError}
     * INTERNAL_SERVER_ERROR if the database operation fails.
     * @throws {TRPCError}
     * NOT_FOUND if the draft does not exist.
     */
    getDraft: protectedProcedure.input(z.object({ formId: z.string() })).query(
        async (
            opts
        ): Promise<{
            formId: string;
            userId: string;
            data: z.infer<typeof draftDataSchema>;
        }> => {
            if (
                !opts.ctx.session.user.roleTitles.includes(
                    UserRoleTitles.ASSIGNEE
                )
            )
                throw new TRPCError({ code: "FORBIDDEN" });

            let result: Awaited<
                ReturnType<typeof opts.ctx.prisma.draft.findUnique>
            >;
            try {
                result = await opts.ctx.prisma.draft.findUnique({
                    where: {
                        formId_userId: {
                            formId: opts.input.formId,
                            userId: opts.ctx.session.user.data.id,
                        },
                    },
                });
            } catch (e) {
                throw new TRPCError({ code: "INTERNAL_SERVER_ERROR" });
            }
            if (result === null) throw new TRPCError({ code: "NOT_FOUND" });

            return {
                ...result,
                data: JSON.parse(result.data) as z.infer<
                    typeof draftDataSchema
                >,
            };
        }
    ),
    /**
     * Delete a draft by form id.
     * @param string - Input.formId the id of the form to delete the draft for.
     * @throws {TRPCError}
     * FORBIDDEN if the user is not a {@link UserRoleTitles.ASSIGNEE}.
     * @throws {TRPCError}
     * INTERNAL_SERVER_ERROR if the database operation fails.
     * @throws {TRPCError}
     * NOT_FOUND if the draft does not exist.
     */
    deleteDraft: protectedProcedure
        .input(z.object({ formId: z.string() }))
        .mutation(async (opts): Promise<Draft> => {
            if (
                !opts.ctx.session.user.roleTitles.includes(
                    UserRoleTitles.ASSIGNEE
                )
            )
                throw new TRPCError({ code: "FORBIDDEN" });

            try {
                return await opts.ctx.prisma.draft.delete({
                    where: {
                        formId_userId: {
                            formId: opts.input.formId,
                            userId: opts.ctx.session.user.data.id,
                        },
                    },
                });
            } catch (e) {
                if (e instanceof Prisma.PrismaClientKnownRequestError) {
                    if (e.code === notFoundErrorCode)
                        throw new TRPCError({ code: "NOT_FOUND" });
                }
                throw new TRPCError({ code: "INTERNAL_SERVER_ERROR" });
            }
        }),
});

export default draftRouter;
