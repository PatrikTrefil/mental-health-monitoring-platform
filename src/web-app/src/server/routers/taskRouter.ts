import { loadFormById } from "@/client/formManagementClient";
import { loadClientsAndPatients } from "@/client/userManagementClient";
import UserRoleTitles from "@/constants/userRoleTitles";
import { Prisma, type Deadline, type Task } from "@prisma/client";
import { TRPCError } from "@trpc/server";
import { z } from "zod";
import { prisma } from "../db";
import { createTRPCRouter, protectedProcedure } from "../trpc";

type TaskWithDeadline = Task & { deadline: Deadline | null };

const taskRouter = createTRPCRouter({
    /**
     * Get list of all tasks. If user is {@link UserRoleTitles.ZADAVATEL_DOTAZNIKU}, return all tasks.
     * If user is {@link UserRoleTitles.KLIENT_PACIENT}, return only tasks for them.
     */
    listTasks: protectedProcedure
        .input(
            z.object({
                limit: z.number().nonnegative(),
                offset: z.number().nonnegative(),
            })
        )
        .query(
            async (
                opts
            ): Promise<{ data: TaskWithDeadline[]; count: number }> => {
                const userRoleTitles = opts.ctx.session.user.roleTitles;
                if (
                    userRoleTitles.includes(UserRoleTitles.ZADAVATEL_DOTAZNIKU)
                ) {
                    const [data, count] = await opts.ctx.prisma.$transaction([
                        opts.ctx.prisma.task.findMany({
                            include: { deadline: true },
                            skip: opts.input.offset,
                            take: opts.input.limit,
                        }),
                        opts.ctx.prisma.task.count(),
                    ]);
                    return { data, count };
                } else if (
                    userRoleTitles.includes(UserRoleTitles.KLIENT_PACIENT)
                ) {
                    const [data, count] = await opts.ctx.prisma.$transaction([
                        opts.ctx.prisma.task.findMany({
                            where: {
                                forUserId: opts.ctx.session.user.data.id,
                            },
                            include: {
                                deadline: true,
                            },
                            skip: opts.input.offset,
                            take: opts.input.limit,
                        }),
                        opts.ctx.prisma.task.count({
                            where: {
                                forUserId: opts.ctx.session.user.data.id,
                            },
                        }),
                    ]);
                    return { data, count };
                } else throw new TRPCError({ code: "FORBIDDEN" });
            }
        ),
    /**
     * Get task by id.
     * @returns The task with given id.
     * @throws {TRPCError}
     * FORBIDDEN if the user is not a {@link UserRoleTitles.ZADAVATEL_DOTAZNIKU} or {@link UserRoleTitles.KLIENT_PACIENT}.
     * @throws {TRPCError}
     * FORBIDDEN if the user is not an {@link UserRoleTitles.ZADAVATEL_DOTAZNIKU} and the task is not assigned to them.
     * @throws {TRPCError}
     * NOT_FOUND if the task does not exist.
     */
    getTask: protectedProcedure
        .input(
            z.object({
                id: z.string(),
            })
        )
        .query(async (opts): Promise<TaskWithDeadline> => {
            const userRoleTitles = opts.ctx.session.user.roleTitles;

            if (
                !userRoleTitles.includes(UserRoleTitles.ZADAVATEL_DOTAZNIKU) &&
                !userRoleTitles.includes(UserRoleTitles.KLIENT_PACIENT)
            )
                throw new TRPCError({ code: "FORBIDDEN" });

            const result = await opts.ctx.prisma.task.findUnique({
                where: {
                    id: opts.input.id,
                },
                include: {
                    deadline: true,
                },
            });
            if (result === null) throw new TRPCError({ code: "NOT_FOUND" });

            if (
                result.forUserId !== opts.ctx.session.user.data.id &&
                !userRoleTitles.includes(UserRoleTitles.ZADAVATEL_DOTAZNIKU)
            ) {
                throw new TRPCError({ code: "FORBIDDEN" });
            }
            return result;
        }),
    /**
     * Create new task.
     * @returns The created task.
     * @throws {TRPCError}
     * FORBIDDEN if the user is not an {@link UserRoleTitles.ZADAVATEL_DOTAZNIKU}.
     * @throws {TRPCError}
     * CONFLICT if the form with given formId does not exist.
     * @throws {TRPCError}
     * CONFLICT if the user with given forUserId does not exist.
     * @throws {TRPCError}
     * INTERNAL_SERVER_ERROR if the checking of form existence fails.
     * @throws {TRPCError}
     * INTERNAL_SERVER_ERROR if the creation of task fails.
     * @throws {TRPCError}
     * INTERNAL_SERVER_ERROR if the loading of users fails.
     */
    createTask: protectedProcedure
        .input(
            z.object({
                name: z.string(),
                forUserId: z.string().nonempty(),
                description: z.string().optional(),
                formId: z.string().nonempty(),
                deadline: z
                    .object({
                        dueDateTime: z.date(),
                        canBeCompletedAfterDeadline: z.boolean(),
                    })
                    .optional(),
                start: z.date().optional(),
            })
        )
        .mutation(async (opts): Promise<TaskWithDeadline> => {
            if (
                !opts.ctx.session.user.roleTitles.includes(
                    UserRoleTitles.ZADAVATEL_DOTAZNIKU
                )
            )
                throw new TRPCError({ code: "FORBIDDEN" });

            let loadedForm:
                | Awaited<ReturnType<typeof loadFormById>>
                | undefined;
            try {
                loadedForm = await loadFormById(
                    opts.input.formId,
                    opts.ctx.session.user.formioToken
                );
            } catch (e) {
                console.error(e);
                throw new TRPCError({
                    code: "INTERNAL_SERVER_ERROR",
                    message: "Checking of form existence failed",
                });
            }

            if (loadedForm === null)
                throw new TRPCError({
                    code: "CONFLICT",
                    message: "Form with given formId does not exist",
                });

            const users = await loadClientsAndPatients(
                opts.ctx.session.user.formioToken
            );

            if (
                users.find((u) => u.data.id === opts.input.forUserId) ===
                undefined
            )
                throw new TRPCError({
                    code: "CONFLICT",
                    message: "User with given forUserId does not exist",
                });

            return opts.ctx.prisma.task.create({
                data: {
                    name: opts.input.name,
                    createdByEmployeeId: opts.ctx.session.user.data.id,
                    forUserId: opts.input.forUserId,
                    description: opts.input.description,
                    formId: opts.input.formId,
                    deadline: { create: opts.input.deadline },
                    start: opts.input.start,
                },
                include: { deadline: true },
            });
        }),
    /**
     * Delete task by id.
     * @throws {TRPCError}
     * FORBIDDEN if the user is not an {@link UserRoleTitles.ZADAVATEL_DOTAZNIKU}.
     * @throws {TRPCError}
     * NOT_FOUND if the task with given id does not exist.
     */
    deleteTask: protectedProcedure
        .input(
            z.object({
                id: z.string(),
            })
        )
        .mutation(async (opts): Promise<void> => {
            if (
                !opts.ctx.session.user.roleTitles.includes(
                    UserRoleTitles.ZADAVATEL_DOTAZNIKU
                )
            )
                throw new TRPCError({ code: "FORBIDDEN" });

            try {
                // https://www.prisma.io/docs/concepts/components/prisma-client/crud#cascading-deletes-deleting-related-records
                await prisma.$transaction([
                    // deleteMany will not fail even if there is nothing to delete (i.e. the task does not have a deadline)
                    opts.ctx.prisma.deadline.deleteMany({
                        where: {
                            taskId: opts.input.id,
                        },
                    }),
                    opts.ctx.prisma.task.delete({
                        where: {
                            id: opts.input.id,
                        },
                    }),
                ]);
            } catch (e) {
                if (
                    e instanceof Prisma.PrismaClientKnownRequestError &&
                    e.code === "P2025" // Record to delete does not exist
                ) {
                    throw new TRPCError({ code: "NOT_FOUND" });
                }
                throw e;
            }
        }),
});

export default taskRouter;
