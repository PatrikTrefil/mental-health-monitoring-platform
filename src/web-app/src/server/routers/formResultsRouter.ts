import {
    loadFormById,
    loadSubmission,
    loadSubmissions,
} from "@/client/formManagementClient";
import UserRoleTitles from "@/constants/userRoleTitles";
import { Submission } from "@/types/formManagement/submission";
import { Prisma, TaskState } from "@prisma/client";
import { TRPCError } from "@trpc/server";
import { Session } from "next-auth";
import { z } from "zod";
import { prisma } from "../db";
import { createTRPCRouter, protectedProcedure } from "../trpc";
import { TaskWithDeadline } from "./taskRouter";

const formResultsRouter = createTRPCRouter({
    /**
     * Form results data joined with task data.
     * @throws {TRPCError} - If the user does not have any of the required roles - {@link UserRoleTitles.ASSIGNER}, {@link UserRoleTitles.FORM_MANAGER}.
     * @throws {TRPCError} - If the form with given ID does not exist.
     */
    getFormResults: protectedProcedure
        .input(
            z.object({
                formId: z.string(),
                sort: z
                    .object({
                        /**
                         * Field to sort by. If the field is nested withing the task object, use dot notation.
                         * @example "name"
                         * @example "deadline.dueDateTime"
                         */
                        field: z.string(),
                        /**
                         * Sort order. Can be either ascending or descending.
                         */
                        order: z.enum(["asc", "desc"]).default("asc"),
                    })
                    .optional(),
                filter: z
                    .object({
                        fieldPath: z.string(),
                        // Possibly add more operators in the future.
                        operation: z.enum(["contains"]),
                        comparedValue: z.string(),
                    })
                    .optional(),
            })
        )
        .query(
            async (
                opts
            ): Promise<{
                data: {
                    task: TaskWithDeadline | null;
                    submission: Submission | null;
                }[];
                totalCount: number;
            }> => {
                validatePermissions(opts.ctx.session);

                let dataToReturn = await getJoinedTasksAndSubmissions(
                    opts.input.formId,
                    opts.ctx.session.user.formioToken,
                    opts.input.filter
                );

                if (opts.input.sort !== undefined) {
                    sortByField(
                        dataToReturn.data,
                        opts.input.sort!.field,
                        opts.input.sort!.order === "desc"
                    );
                }
                return dataToReturn;
            }
        ),
});

export default formResultsRouter;

/**
 * Get tasks and submissions data joined on id of the form they are associated with.
 * @param formId - The ID of the form to load data for.
 * @param token - The token to use to authenticate with the form management API.
 * @param filter - The filter to apply to the submissions.
 * @param filter.comparedValue - The value to compare the field to.
 * @param filter.fieldPath - The path of the field to compare.
 * @param filter.operation - The operation to use to compare the field to the compared value.
 * @returns Unsorted joined data about tasks and submissions for a particular form.
 */
async function getJoinedTasksAndSubmissions(
    formId: string,
    token: string,
    filter?: { comparedValue: string; fieldPath: string; operation: "contains" }
): Promise<{
    data: {
        submission: Submission | null;
        task: TaskWithDeadline | null;
    }[];
    totalCount: number;
}> {
    if (filter && filter.fieldPath.startsWith("submission")) {
        const { submissions, totalCount } = await loadAllSubmissions(
            formId,
            token,
            {
                comparedValue: filter.comparedValue,
                fieldPath: filter.fieldPath.slice("submission.".length),
                operation: filter.operation,
            }
        );
        let taskPromises: Promise<TaskWithDeadline | null>[] = [];
        for (const submission of submissions) {
            taskPromises.push(
                prisma.task.findFirst({
                    where: {
                        submissionId: submission._id,
                    },
                    include: {
                        deadline: true,
                    },
                })
            );
        }
        const tasks = await Promise.allSettled(taskPromises);
        return {
            data: tasks.map((task, i) => {
                if (task.status === "fulfilled") {
                    return {
                        submission: submissions[i]!,
                        task: task.value!,
                    };
                } else {
                    return {
                        submission: submissions[i]!,
                        task: null,
                    };
                }
            }),
            totalCount,
        };
    } else {
        const { tasks, count } = await loadAllTasksForForm(formId, filter);
        if (tasks === undefined) {
            return { data: [], totalCount: 0 };
        } else {
            const submissionPromises: Promise<Submission | null>[] = [];
            const form = await loadFormById(formId, token);
            if (form === null) throw new TRPCError({ code: "NOT_FOUND" });
            for (const task of tasks) {
                if (task.submissionId)
                    submissionPromises.push(
                        loadSubmission(form?.path, task.submissionId, token)
                    );
            }
            const submissions = await Promise.allSettled(submissionPromises);
            return {
                data: submissions.map((submission, i) => {
                    if (submission.status === "fulfilled") {
                        return {
                            submission: submission.value!,
                            task: tasks[i]!,
                        };
                    } else {
                        return {
                            submission: null,
                            task: tasks[i]!,
                        };
                    }
                }),
                totalCount: count,
            };
        }
    }
}

/**
 * Validate that the user has the required permissions to access the form results.
 * @param session - The session of the user.
 * @throws {TRPCError} - If the user does not have any of the required roles - {@link UserRoleTitles.ASSIGNER}, {@link UserRoleTitles.FORM_MANAGER}.
 */
function validatePermissions(session: Session) {
    if (
        !(
            session.user.roleTitles.includes(UserRoleTitles.FORM_MANAGER) ||
            session.user.roleTitles.includes(UserRoleTitles.ASSIGNER)
        )
    )
        throw new TRPCError({ code: "FORBIDDEN" });
}

/**
 * Load all submissions that match the given filter. This function does not support pagination.
 * @param formId - The ID of the form to load submissions from.
 * @param token - The token to use to authenticate with the form management API.
 * @param filter - The filter to apply to the submissions.
 * @param filter.comparedValue - The value to compare the field to.
 * @param filter.fieldPath - The path of the field to compare.
 * @param filter.operation - The operation to use to compare the field to the compared value.
 */
async function loadAllSubmissions(
    formId: string,
    token: string,
    filter?: { comparedValue: string; fieldPath: string; operation: "contains" }
): Promise<{ submissions: Submission[]; totalCount: number }> {
    const submissions: Submission[] = [];
    let totalCount = 1; // Init to 1 so that the loop runs at least once (overwritten at the end of every iteration).
    while (totalCount > submissions.length) {
        const currResult = await loadSubmissions(formId, {
            token,
            pagination: {
                limit: 100,
                offset: submissions.length,
            },
            filters: filter !== undefined ? [filter] : undefined,
        });
        submissions.push(...currResult.data);
        totalCount = currResult.totalCount;
    }
    return { submissions, totalCount };
}

/**
 * Load all tasks for a form that match the given filter.
 * @param formId - The ID of the form to load tasks for.
 * @param filter - The filter to apply to the tasks.
 * @param filter.comparedValue - The value to compare the field to.
 * @param filter.fieldPath - The path of the field to compare.
 * @param filter.operation - The operation to use to compare the field to the compared value.
 */
async function loadAllTasksForForm(
    formId: string,
    filter?: { comparedValue: string; fieldPath: string; operation: "contains" }
): Promise<{ tasks: TaskWithDeadline[]; count: number }> {
    const filters: Prisma.TaskWhereInput["AND"] = [];
    if (filter && filter.fieldPath.startsWith("task")) {
        const taskPrefix = "task.";
        const deadlinePrefix = "deadline.";
        if (filter.fieldPath.startsWith(deadlinePrefix)) {
            filters.push({
                deadline: {
                    [filter.fieldPath.slice(
                        taskPrefix.length + deadlinePrefix.length
                    )]: {
                        contains: filter.comparedValue,
                    },
                },
            });
        } else {
            filters.push({
                [filter.fieldPath.slice(taskPrefix.length)]: {
                    contains: filter.comparedValue,
                },
            });
        }
    }
    const [tasks, count] = await prisma.$transaction([
        prisma.task.findMany({
            where: {
                formId: formId,
                AND: {
                    state: TaskState.COMPLETED,
                    AND: filters,
                },
            },
            include: {
                deadline: true,
            },
        }),
        prisma.task.count({
            where: {
                formId: formId,
                AND: {
                    state: TaskState.COMPLETED,
                    AND: filters,
                },
            },
        }),
    ]);
    return { tasks, count };
}

/**
 * Sort data in place by field, which is defined by a dot-separated path.
 * The given data array is mutated.
 * Values are considered equal if they do not exist on the objects.
 * Values are considered equal if they are not a number or string.
 * @param data - The data to sort.
 * @param fieldPathStr - The path of the field to sort by.
 * @param reverse - Whether to sort in reverse order.
 * @returns The sorted data (reference to the original array, which has been mutated).
 */
function sortByField(data: any[], fieldPathStr: string, reverse = false) {
    const fieldPath = fieldPathStr.split(".");
    data.sort((a, b) => {
        let aValue: any = a;
        let bValue: any = b;
        for (const key of fieldPath) {
            if (
                typeof aValue === "object" &&
                typeof bValue === "object" &&
                aValue !== null &&
                bValue !== null &&
                key in aValue &&
                key in bValue
            ) {
                aValue = aValue[key];
                bValue = bValue[key];
            } else {
                return 0;
            }
        }
        if (typeof aValue === "string" && typeof bValue === "string") {
            return aValue.localeCompare(bValue);
        } else if (typeof aValue === "number" && typeof bValue === "number") {
            return aValue - bValue;
        } else {
            return 0;
        }
    });
    if (reverse) data.reverse();
    return data;
}
