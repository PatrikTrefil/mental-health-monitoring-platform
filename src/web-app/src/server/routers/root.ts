import { createTRPCRouter } from "../trpc";
import draftRouter from "./draftRouter";
import formResultsRouter from "./formResultsRouter";
import taskRouter from "./taskRouter";

export const appRouter = createTRPCRouter({
    task: taskRouter,
    draft: draftRouter,
    formResults: formResultsRouter,
});

// export type definition of API
export type AppRouter = typeof appRouter;
