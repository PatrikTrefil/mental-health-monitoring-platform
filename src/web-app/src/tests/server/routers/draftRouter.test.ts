import UserRoleTitles from "@/constants/userRoleTitles";
import { AppRouter, appRouter } from "@/server/routers/root";
import { faker } from "@faker-js/faker";
import { inferProcedureInput } from "@trpc/server";
import { describe, expect, it } from "vitest";
import { createInnerTRPCContextMockSession } from "./utility";

// make tests deterministic
faker.seed(123);

describe.each(
    Object.values(UserRoleTitles).filter(
        (r) => r !== UserRoleTitles.KLIENT_PACIENT
    )
)("as $1 it", (role) => {
    const ctx = createInnerTRPCContextMockSession([role]);
    const caller = appRouter.createCaller(ctx);

    it("should not be possible to create a draft", async () => {
        type CreateDraftInput = inferProcedureInput<
            AppRouter["draft"]["upsertDraft"]
        >;
        const input: CreateDraftInput = {
            formId: "some-form-id",
            data: {},
        };
        expect(
            async () => await caller.draft.upsertDraft(input)
        ).rejects.toMatchSnapshot();
    });

    it("should not be possible to get a draft", async () => {
        expect(
            async () => await caller.draft.getDraft({ formId: "some-form-id" })
        ).rejects.toMatchSnapshot();
    });

    it("should not be possible to update a draft", async () => {
        expect(
            async () =>
                await caller.draft.upsertDraft({
                    formId: "some-form-id",
                    data: {},
                })
        ).rejects.toMatchSnapshot();
    });
});

describe(`as a ${UserRoleTitles.KLIENT_PACIENT} it`, () => {
    const ctx = createInnerTRPCContextMockSession([
        UserRoleTitles.KLIENT_PACIENT,
    ]);
    const caller = appRouter.createCaller(ctx);
    const nonexistentFormId = "nonexistent-form-id";

    it("throws when getting a non-existing draft", async () => {
        expect(
            caller.draft.getDraft({ formId: nonexistentFormId })
        ).rejects.toMatchSnapshot();
    });

    it("throws when deleting a non-existing draft", async () => {
        expect(
            caller.draft.deleteDraft({ formId: nonexistentFormId })
        ).rejects.toMatchSnapshot();
    });

    it("should throw when passing a non-stringifiable JSON value", async () => {
        expect(
            caller.draft.upsertDraft({
                formId: nonexistentFormId,
                data: {
                    a: BigInt(10),
                },
            })
        ).rejects.toMatchSnapshot();
    });

    const formId = faker.string.alpha(10);

    it("should be possible to create a draft", async () => {
        type CreateDraftInput = inferProcedureInput<
            AppRouter["draft"]["upsertDraft"]
        >;
        const input: CreateDraftInput = {
            formId: formId,
            data: {},
        };
        const result = await caller.draft.upsertDraft(input);
        expect(result).toMatchSnapshot();
    });

    it("should be possible to edit my draft", async () => {
        type UpdateDraftInput = inferProcedureInput<
            AppRouter["draft"]["upsertDraft"]
        >;
        const input: UpdateDraftInput = {
            formId: formId,
            data: { new: "data" },
        };
        const result = await caller.draft.upsertDraft(input);
        expect(result).toMatchSnapshot();
    });

    it("should be possible to delete a draft ", async () => {
        await caller.draft.deleteDraft({ formId: formId });
    });
});
