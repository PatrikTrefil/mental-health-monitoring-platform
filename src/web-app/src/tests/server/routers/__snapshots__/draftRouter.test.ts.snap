// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`as $1 it > should not be possible to create a draft 1`] = `[TRPCError: FORBIDDEN]`;

exports[`as $1 it > should not be possible to create a draft 2`] = `[TRPCError: FORBIDDEN]`;

exports[`as $1 it > should not be possible to create a draft 3`] = `[TRPCError: FORBIDDEN]`;

exports[`as $1 it > should not be possible to create a draft 4`] = `[TRPCError: FORBIDDEN]`;

exports[`as $1 it > should not be possible to get a draft 1`] = `[TRPCError: FORBIDDEN]`;

exports[`as $1 it > should not be possible to get a draft 2`] = `[TRPCError: FORBIDDEN]`;

exports[`as $1 it > should not be possible to get a draft 3`] = `[TRPCError: FORBIDDEN]`;

exports[`as $1 it > should not be possible to get a draft 4`] = `[TRPCError: FORBIDDEN]`;

exports[`as $1 it > should not be possible to update a draft 1`] = `[TRPCError: FORBIDDEN]`;

exports[`as $1 it > should not be possible to update a draft 2`] = `[TRPCError: FORBIDDEN]`;

exports[`as $1 it > should not be possible to update a draft 3`] = `[TRPCError: FORBIDDEN]`;

exports[`as $1 it > should not be possible to update a draft 4`] = `[TRPCError: FORBIDDEN]`;

exports[`as a Klient/Pacient it > should be possible to create a draft 1`] = `
{
  "data": "{}",
  "formId": "some-form-id",
  "userId": "123",
}
`;

exports[`as a Klient/Pacient it > should be possible to edit my draft 1`] = `
{
  "data": "{\\"new\\":\\"data\\"}",
  "formId": "some-form-id",
  "userId": "123",
}
`;

exports[`as a Klient/Pacient it > should throw when passing a non-stringifiable JSON value 1`] = `
[TRPCError: [
  {
    "message": "Data can not be stringified to JSON",
    "code": "custom",
    "path": [
      "data"
    ]
  }
]]
`;

exports[`as a Klient/Pacient it > throws when deleting a non-existing draft 1`] = `[TRPCError: NOT_FOUND]`;

exports[`as a Klient/Pacient it > throws when getting a non-existing draft 1`] = `[TRPCError: NOT_FOUND]`;
